From: liuchang0812 <liuchang0812@gmail.com>
Date: Fri, 30 Jun 2017 16:50:53 +0800
Subject: os: export compact interface in ObjectStore and ObjectMap

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>
(cherry picked from commit eb5723d3abd4e1b6caec477090761d7919efc7bd)

 Conflicts:
	src/os/bluestore/BlueStore.h
         Removed declarations which are not part of this backport
          inject_data_error()
          inject_mdata_error()
          debug_data_eio()
          debug_mdata_eio()
          debug_oj_on_delete()
	src/os/filestore/FileStore.h
         Removed declarations which are not part of this backport
          set<ghobject_t> data_error_set
          set<ghobject_t> mdata_error_set
          inject_data_error() override
          inject_mdata_error() override

(cherry picked from commit 9de0f8f68aab7e90b19e97570fbc0075204231d8)

Resolves: rhbz#1483224
---
 src/os/ObjectMap.h             | 2 ++
 src/os/ObjectStore.h           | 2 ++
 src/os/bluestore/BlueStore.h   | 5 +++++
 src/os/filestore/DBObjectMap.h | 5 +++++
 src/os/filestore/FileStore.cc  | 1 +
 src/os/filestore/FileStore.h   | 6 ++++++
 src/os/kstore/KStore.h         | 5 +++++
 7 files changed, 26 insertions(+)

diff --git a/src/os/ObjectMap.h b/src/os/ObjectMap.h
index c4efc7f..ddffdb6 100644
--- a/src/os/ObjectMap.h
+++ b/src/os/ObjectMap.h
@@ -139,6 +139,8 @@ public:
 
   virtual bool check(std::ostream &out) { return true; }
 
+  virtual void compact() {}
+
   typedef KeyValueDB::GenericIteratorImpl ObjectMapIteratorImpl;
   typedef ceph::shared_ptr<ObjectMapIteratorImpl> ObjectMapIterator;
   virtual ObjectMapIterator get_iterator(const ghobject_t &oid) {
diff --git a/src/os/ObjectStore.h b/src/os/ObjectStore.h
index 93ae4bb..cc3c9a1 100644
--- a/src/os/ObjectStore.h
+++ b/src/os/ObjectStore.h
@@ -2409,6 +2409,8 @@ public:
   // DEBUG
   virtual void inject_data_error(const ghobject_t &oid) {}
   virtual void inject_mdata_error(const ghobject_t &oid) {}
+
+  virtual void compact() {}
 };
 WRITE_CLASS_ENCODER(ObjectStore::Transaction)
 WRITE_CLASS_ENCODER(ObjectStore::Transaction::TransactionData)
diff --git a/src/os/bluestore/BlueStore.h b/src/os/bluestore/BlueStore.h
index 91aec62..c2c577f 100644
--- a/src/os/bluestore/BlueStore.h
+++ b/src/os/bluestore/BlueStore.h
@@ -840,6 +840,11 @@ public:
     TrackedOpRef op = TrackedOpRef(),
     ThreadPool::TPHandle *handle = NULL) override;
 
+  void compact() override {
+    assert(db);
+    db->compact();
+  }
+  
 private:
   // --------------------------------------------------------
   // write ops
diff --git a/src/os/filestore/DBObjectMap.h b/src/os/filestore/DBObjectMap.h
index de68f3c..aace89f 100644
--- a/src/os/filestore/DBObjectMap.h
+++ b/src/os/filestore/DBObjectMap.h
@@ -217,6 +217,11 @@ public:
   /// Ensure that all previous operations are durable
   int sync(const ghobject_t *oid=0, const SequencerPosition *spos=0);
 
+  void compact() override {
+    assert(db);
+    db->compact();
+  }
+
   /// Util, list all objects, there must be no other concurrent access
   int list_objects(vector<ghobject_t> *objs ///< [out] objects
     );
diff --git a/src/os/filestore/FileStore.cc b/src/os/filestore/FileStore.cc
index d2e969f..ac98f8b 100644
--- a/src/os/filestore/FileStore.cc
+++ b/src/os/filestore/FileStore.cc
@@ -4107,6 +4107,7 @@ void FileStore::inject_mdata_error(const ghobject_t &oid) {
   dout(10) << __func__ << ": init error on " << oid << dendl;
   mdata_error_set.insert(oid);
 }
+
 void FileStore::debug_obj_on_delete(const ghobject_t &oid) {
   Mutex::Locker l(read_error_lock);
   dout(10) << __func__ << ": clear error on " << oid << dendl;
diff --git a/src/os/filestore/FileStore.h b/src/os/filestore/FileStore.h
index 9d3f9c8..7d792c4 100644
--- a/src/os/filestore/FileStore.h
+++ b/src/os/filestore/FileStore.h
@@ -595,6 +595,12 @@ public:
   set<ghobject_t, ghobject_t::BitwiseComparator> mdata_error_set; // getattr(),stat() will return -EIO
   void inject_data_error(const ghobject_t &oid);
   void inject_mdata_error(const ghobject_t &oid);
+
+  void compact() override {
+    assert(object_map);
+    object_map->compact();
+  }
+
   void debug_obj_on_delete(const ghobject_t &oid);
   bool debug_data_eio(const ghobject_t &oid);
   bool debug_mdata_eio(const ghobject_t &oid);
diff --git a/src/os/kstore/KStore.h b/src/os/kstore/KStore.h
index 09483de..c27df7e 100644
--- a/src/os/kstore/KStore.h
+++ b/src/os/kstore/KStore.h
@@ -532,6 +532,11 @@ public:
     TrackedOpRef op = TrackedOpRef(),
     ThreadPool::TPHandle *handle = NULL);
 
+  void compact () override {
+    assert(db);
+    db->compact();
+  }
+  
 private:
   // --------------------------------------------------------
   // write ops
